pkg/autoscaler/health.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package autoscaler
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/golang/glog"
+)
+
+type healthInfo struct {
+	m   *sync.RWMutex
+	err error
+}
+
+func newHealthInfo() *healthInfo {
+	return &healthInfo{m: &sync.RWMutex{}, err: nil}
+}
+
+func (h *healthInfo) setError(err error) {
+	h.m.Lock()
+	h.err = err
+	h.m.Unlock()
+}
+
+func (h *healthInfo) getError() error {
+	h.m.RLock()
+	defer h.m.RUnlock()
+	return h.err
+}
+
+func (s *AutoScaler) startHealthz() {
+	http.HandleFunc("/healthz", s.healthFn)
+	glog.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func (s *AutoScaler) healthFn(w http.ResponseWriter, req *http.Request) {
+	if req.Header.Get("CheckCachedError") != "true" {
```

Comment:
I think this is overkill.  

If you want the status of the last poll, OK. Let's expose that.  We don't need a header to decide that.  Just use a different URL path.

/healthz -> is it alive?
/last-poll -> did it sync on the last cycle?

How you use that info is what is interesting.  If the poll cycle and the probe cycle are not configured well, you could end up failing the probe "repeatedly" without having actually polled the server in between.

An alternative approach could be to take a flag indicating how many failures to allow and then exit().

Reply:
Thanks for the suggestions!

What do you think about keeping the two endpoints as well as having your suggested alternative (flag to force restart after maxAttempts)? I don't believe they will step over each other. I've pushed code to include both. 

Also did a bit of a refactor to have the health server as a dependency.